src/include/duckdb/common/types/string.hpp

```diff
@@ -0,0 +1,270 @@
+		return (len & NON_OWN_BIT) == 0;
+	}
+	bool IsInline() const {
+		return len <= INLINE_MAX;
+	}
+	static bool CanBeInlined(size_t size) {
+		return size <= INLINE_MAX;
+	}
+
+	// Creates a new String instance with its own copy of the data
+	static String Copy(const char *data, size_t size) {
+		String result;
+		result.AssignOwning(data, size);
+		return result;
+	}
+
+	static String Copy(const String &other) {
+		return Copy(other.data(), other.size());
+	}
+	static String Copy(const char *data) {
+		return Copy(data, data ? strlen(data) : 0);
+	}
+	static String Copy(const std::string &str) {
+		return Copy(str.data(), str.size());
+	}
+	String Copy() const {
+		return Copy(data(), size());
+	}
+
+	// Creates a new String instance that references the data without owning it
+	// If the size is small enough, it will inline the data; which WILL be owning
+	static String Reference(const char *data, size_t size) {
+		String result;
+		// If we reference, and we can inline it, we make owning anyway
+		if (size <= INLINE_MAX) {
+			result.AssignOwning(data, size);
+		} else {
+			result.ptr = const_cast<char *>(data); // NOLINT allow const cast
+			result.len = size | NON_OWN_BIT;       // Set the non-owning bit
+		}
+		return result;
+	}
+
+	static String Reference(const String &other) {
+		return Reference(other.data(), other.size());
+	}
+	static String Reference(const char *data) {
+		return Reference(data, data ? strlen(data) : 0);
+	}
+	static String Reference(const std::string &str) {
+		return Reference(str.data(), str.size());
+	}
+	String Reference() const {
+		return Reference(data(), size());
+	}
+
+	string ToStdString() const {
+		if (IsInline()) {
+			return string(buf, size());
+		}
+		return string(ptr, size());
+	}
+
+public:
+	static char CharacterToLower(char c) {
+		if (c >= 'A' && c <= 'Z') {
+			return UnsafeNumericCast<char>(c + ('a' - 'A'));
+		}
+		return c;
+	}
+
+	String Lower() const {
+		const auto str_data = data();
+		const auto str_size = size();
+
+		std::string lowercase_str;
+		lowercase_str.reserve(str_size);
+		for (idx_t i = 0; i < str_size; ++i) {
+			lowercase_str.push_back(CharacterToLower(str_data[i]));
+		}
+		return String(lowercase_str);
+	}
+
+private:
+	static constexpr auto INLINE_CAP = sizeof(char *);
+	static constexpr auto INLINE_MAX = INLINE_CAP - 1;
+	static constexpr auto NON_OWN_BIT = 1UL << (sizeof(size_t) * 8 - 1);
+	static constexpr auto LENGTH_MAX = (1UL << (sizeof(size_t) * 8 - 1)) - 1;
+
+	void AssignOwning(const char *new_data, size_t new_size) {
+		len = new_size;
+
+		if (len == 0) {
+			buf[len] = '\0'; // Null-terminate the inline buffer
+			return;
+		}
+
+		D_ASSERT(new_data != nullptr);
+
+		if (len <= INLINE_MAX) {
+			memcpy(buf, new_data, len);
+			buf[len] = '\0'; // Null-terminate the inline buffer
+			return;
+		}
+
+		auto new_ptr = new char[len + 1]; // +1 for null-termination
+		memcpy(new_ptr, new_data, len);
+		new_ptr[len] = '\0';
+
+		ptr = new_ptr;
```

Comment:
It's this currently ever triggered?

I think it's a bit weird that interface misuse might lead to actual allocations happening without explicit control.
Unsure how to solve this, should there be an ifdef that trigger an exception if this ever happen?

Reply:
What is the misuse here? AssignOwning is only called internally and semantically constructs an owning string by copy from some data.